WIP: impl `derp::client` mod

ramfox opened the pull request:
some general notes about this first attempt (aka "this mess"):

- golang impl uses u32 everywhere, including sending the frame length on the wire.  I assuming we should continue to only use 4 bytes for the frame length, but then we should probably convert to `usize` when we get it, so that we have fewer `as usize`'s littered throughout the code. But I haven't done this yet.

- from looking at the surrounding code that imports the derp client, it looks like we need to be able to use this in many threads. So I've wrapped any "shared data" in an `Arc<Mutex>`

- I don't understand why the go code only wrapped the `writer` in a mutex lock and not the `reader` as well. I followed the same pattern, because maybe I don't understand this yet?

- the rate limiter is gross and maybe we wrap this in a trait so we don't have to drag around the underlying `governor` traits in our client implementation. Updating the rate limiter is also not fully implemented. Because I'm mad at how gross it is & also I needed to stop spending time on it & move on.

- this `recv_err` nonsense is so ugly, god help me. After we encounter an error reading while running `recv`, the go impl wants us to return the same error anytime we call `recv` again on the client. (The client is "dead" after the first error). However, most errors in rust aren't clonable, so currently I'm storing the error as a `String` that I've pulled from the error itself. There has got to be a better way to do this, other than the nonsense I've implemented, no?

flub commented on src/hp/derp/client.rs:
```diff
@@ -0,0 +1,515 @@
+            if frame_len as usize > MAX_FRAME_SIZE {
+                bail!("unexpectedly large frame of {} bytes returned", frame_len);
+            }
+            let mut read_total = 0;
+            loop {
+                let read = self.reader.read(&mut frame_payload).await?;
+                if read == 0 {
+                    break;
+                }
+                read_total += read;
+            }
+            if read_total != frame_len as usize {
+                bail!(
+                    "unexpected number of bytes sent in frame, said {}, received {}",
+                    frame_len,
+                    read_total
+                );
+            }
```
aren't you re-implementing `read_exact()`?  if you create the frame_payload after you know frame_len you might be able to use that instead?

ramfox replied:
yes, I did... I basically couldn't figure out how to specify the exact length of the buffer I wanted to use... until just now realizing I could pass a slice of a specific size.

Thanks!

flub commented on src/hp/derp/client.rs:
```diff
@@ -0,0 +1,515 @@
+
+    async fn recv_check_error(&self, timeout: Duration) -> Result<ReceivedMessage> {
+        {
+            let recv_err = self.recv_err.lock().await;
+            if let Some(err) = &*recv_err {
+                bail!(err.clone());
+            }
```
do you really care about the exact error?  We use anyhow errors anyway so i don't think we do.  you can just store an (atomic) bool for the error state for now i think.

ramfox replied:
I... don't know if we care about the exact error. Maybe for now, we go the atomic bool route & just return an anyhow error saying something like "client has encountered a read error and can no longer receive messages"

ramfox commented on src/hp/derp/client.rs:
```diff
@@ -0,0 +1,515 @@
+    }
+
+    async fn parse_server_info(&self, buf: &mut [u8]) -> Result<ServerInfo> {
+        // TODO: should NONCE_LEN just be usize?
```
changed the constants to `usize`

ramfox commented on src/hp/derp/client.rs:
```diff
@@ -0,0 +1,515 @@
+            Some(rl) => Some(Arc::clone(&rl)),
+        };
+        {
+            let mut writer = self.writer.lock().await;
```
okay thank you!

ramfox commented on src/hp/derp/client.rs:
```diff
@@ -0,0 +1,515 @@
+    MW: governor::middleware::RateLimitingMiddleware<C::Instant>,
+{
+    // TODO: for something relatively straight forward, this is pretty hard to follow
+    // TODO: also, should Client.server_key be an option?
```
I'm just going to leave this for now because I think part of the problem is the original design for `read_frame` that the go impl has, and all the reading / writing needs to be looked at more carefully anyway. But thank you for dropping some knowledge here.

ramfox commented:
Okay more notes:
- `const`s are now all `usize`, except the `FrameType`s
- removed all `Mutex`s (including the one on the writer) until rust yells at us
- wrote a `RateLimiter` struct that hides all the ugliness of hanging onto the `governor::RateLimiter`
- cleaned up some `PublicKey` vs `SecretKey` vs `SharedKey` nonsense
- using `AtomicBool` to track if the client is dead

things I'm worried about:
- there are too many ways we attempt to read and write things & I don't think I implemented the most efficient way to do any of them. I hope the surface area of this gets smaller and more correct as we refactor
- we need to talk about the `mesh_key`. In the go impl it's a string. I make it an array because we send the `mesh_key` over the wire & I wanted to use postcard to serialize the struct (and postcard can't estimate the `POSTCARD_MAX_SIZE` with `String` fields.) It was just the quickest way for me to get this done, but we need to figure out the "right" way at some point.

flub reviewed (approved):
there's a couple of blocking bugs wrt to reading from the streams, but i don't want to hold this up once they're fixed

ramfox commented on src/hp/derp.rs:
```diff
@@ -163,23 +163,25 @@ async fn read_frame_header(mut reader: impl AsyncRead + Unpin) -> Result<(FrameT
     }
-
+    // TODO: trusting that the caller has sized the buffer correctly? Shouldn't we just attempt to
+    // read `frame_len` amount?
```
🔥 thank you

ramfox commented on src/hp/derp/client.rs:
```diff
@@ -1,21 +1,532 @@
+        dstkey: PublicKey,
+        packet: &[u8],
+    ) -> Result<()> {
+        tokio::select! {
```
ahh I thought this only really mattered in loops. adding it in.

ramfox commented on src/hp/derp/client.rs:
```diff
@@ -1,21 +1,532 @@
+        // from the tailscale/derp/derp_client.go:
+        // "In practice [a frame is] 4KB (from derphttp.Client's bufio.NewReader(httpConn)) and
+        // in practice, WireGuard packets (and thus DERP frames) are under 1.5 KiB."
+        let mut frame_payload = BytesMut::with_capacity(4 * 1024);
```
Ah, no they don't allocate the 4KB explicitly. I just read this backwards.

We aren't using WireGuard frames tho, we are using quic frames (over HTTP) right? Those seem to be less than 1.5KB as well (https://blog.apnic.net/2019/03/04/a-quick-look-at-quic/#:~:text=QUIC%20and%20IP%20fragmentation&text=For%20IPv4%2C%20the%20QUIC%20maximum,is%201%2C392%20packets%20in%20size), so I'll change this.

flub replied:
QUIC frames could be anything above 1200 bytes.  But QUIC is supposed to find the path MTU so they never get fragmented packets and realistically on the internet this is unlikely to exceed 1500 as that's the standard ethernet frame size.

ramfox commented on src/hp/derp/client.rs:
```diff
@@ -1,21 +1,532 @@
+                bail!("unexpectedly large frame of {} bytes returned", frame_len);
+            }
+            self.reader
+                .read_exact(&mut frame_payload[..frame_len])
```
thank you

ramfox commented:
fixed up the reading and writing story & added read/write round trip tests
